redis/src/tls.rs

```diff
@@ -0,0 +1,207 @@
+use std::io::{BufRead, Error, ErrorKind as IOErrorKind};
+
+use rustls::{Certificate, OwnedTrustAnchor, PrivateKey, RootCertStore};
+use rustls_pemfile::{certs, read_one};
+
+use crate::{
+    Client, ConnectionAddr, ConnectionInfo, ErrorKind, RedisConnectionInfo, RedisError, RedisResult,
+};
+
+/// The TLS client builder type
+pub struct TlsClientBuilder {
+    host: String,
+    port: u16,
+    client_key: Option<PrivateKey>,
+    client_cert: Vec<Certificate>,
+    root_cert: RootCertStore,
+}
+
+/// The builder takes parameters all parameters necessary to create a TLS connection.
+///
+/// It must be created with a _server_ and a _port_ and should be provided with
```

Comment:
if the builder is incomplete without the binary stream, its creation structure shouldn't allow `build` to be called without the stream, so the stream should be passed in constructor, or in `build`. Maybe pass it in an enum denoting the different values, instead of using the builder pattern?

Reply:
yeah, makes sense. 

Since this is a case for tls clients, maybe something like a `TlsParamBuilder` struct that can be "loaded" with keys and certificates, and used to generate a `TlsParam` value if they are consistent.
Then this `TlsParam` structure could  be used to eventually build the client.

Something like 
```
let mut tls_param_builder = TlsParamBuilder::new();

// set CA, client certificates and key
// and handle errors in case of inconsistent data

let tls_param: TlsParam = tls_param_builder.build().expect("parameters have been properly set"); 

let tls_client = redis::Client::open_tls(server, port, tls_param);
```

wdyt?

Reply:
I'm not sure what're the dependencies between the builder functions here. is the builder valid if only one of `client_cert`, `client_key`, or `root_cert` are called? Is it valid if all 3 are called? only some?
Essentially, in the 6 combinatorial options we have for the builder to be prepared before being built - which are valid and which aren't?

I assume that there are dependencies which aren't reflected in the API between the files' contents, and I'm not sure that this can be representable in the API, but if we can be certain that if function A is called, then function B has to / should not be called, we should endeavor to create an API in which it is impossible to not call B / to call B after A.

Reply:
No combinations, mutual TLS needs all 3 parameters: `client_cert`, `client_key`, and `root_cert`.

These 3 parameters are contents of PEM files passed as bytestreams to the builder, and we must ensure they are all valid (convertible into `Certificate`, `PrivateKey` and `RootCertStore` respec) before we can use them to establish a mTLS connection. 

So the idea is to split the API in 2 parts: 

### 1 - Tls parameter creation

Parameter builder provides a way to the user to have these parameters properly validated before proceeding to create the tls_client.

Extending the client code above:
```rust
 let mut tls_param_builder = TlsParamBuilder::new();

// client certificate
 let mut reader = BufReader::new(cert_file);
 tls_param_builder
     .client_cert(&mut reader)
     .expect("Unable to handle client certs");

// client key
 let key_file = File::open(key_file).expect("Cannot open private key file");
 let mut reader = BufReader::new(key_file);
 tls_param_builder
     .client_key(&mut reader)
     .expect("Unable to handle client key");

// root cert
 let ca_file = File::open(ca_file).expect("Cannot open CA cert file");
 let mut reader = BufReader::new(ca_file);
 tls_param_builder
     .root_cert(&mut reader)
     .expect("Unable to load CA cert");

// A
let tls_param: TlsParam = tls_param_builder.build().expect("parameters have not been properly set"); 
```

I feel like enforcing the order of parameters supplied (`cert_file`, then `key_file`, then `root_cert`) would make some scenarios confusing:
_e.g._ if `cert_file` and `key_file` provided are fine, then an inconsistent `root_cert` would return an error and abort the whole builder?

Therefore `TlsParamBuilder`'s `build()` function would return `Ok(TlsParam)` only if the parameters were properly set, and issues during setting the parameters could be handled w/o destroying the builder.

### 2 - TLS client creation

Provided we have valid TLS parameters, user can create the client

```rust
// B
let tls_client = redis::Client::open_tls(server, port, tls_param);
```

Reply:
but this still uses the builder pattern, that implies that the client can be built without any or all of the 3 - you essentially move the check that all 3 were provided to runtime / document reading time. why not dump the builder pattern and enforce that these are required arguments in the function signature?

`Client::build_with_tls(host, port, client_cert, client_key, root_cert)`? This way the user can't try to build without the files, and there's less room for confusion.